doc/macsec-gearbox/PAI-Proposal-Failover.md

```diff
@@ -0,0 +1,381 @@
+
+## Failover Configuration on PHY
+
+Failover feature can be configured on all ports of the PHY. Failover ports can operate in either hitless (the ability to switch over port without losing link) or non-hitless mode.
+
+Additions to saiport.h
+
+/\*\*
+
+\* @brief Attribute data for #SAI\_PORT\_CONNECTOR\_ATTR\_FAILOVER\_MODE
+
+\* Used for Failover mode configuration on port
+
+\*/
+
+typedef enum \_sai\_port\_connector\_failover\_mode\_e
+
+{
+
+/\*\* Failover mode disable \*/
+
+SAI\_PORT\_CONNECTOR\_FAILOVER\_MODE\_DISABLE,
+
+/\*\* Configure Failover mode on primary port \*/
+
+SAI\_PORT\_CONNECTOR\_FAILOVER\_MODE\_PRIMARY,
+
+/\*\* Configure Failover mode on secondary port \*/
+
+SAI\_PORT\_CONNECTOR\_FAILOVER\_MODE\_SECONDARY
+
+} sai\_port\_connector\_failover\_mode\_t;
+
+/\*\*
+
+\* @brief Configure the failover mode on port
+
+\*
+
+\* @type sai\_port\_connector\_failover\_mode\_t
+
+\* @flags CREATE\_AND\_SET
+
+\* @default SAI\_PORT\_CONNECTOR\_FAILOVER\_MODE\_DISABLE
+
+\*/
+
+SAI\_PORT\_CONNECTOR\_ATTR\_FAILOVER\_MODE,
+
+/\*\*
+
+\* @brief System Side Port ID
+
+\*
+
+\* @type sai\_object\_id\_t
+
+\* @flags CREATE\_ONLY
+
+\* @objects SAI\_OBJECT\_TYPE\_PORT
+
+\*/
+
+SAI\_PORT\_CONNECTOR\_ATTR\_SYSTEM\_SIDE\_FAILOVER\_PORT\_ID,
+
+/\*\*
+
+\* @brief Line Side Port ID
+
+\*
+
+\* @type sai\_object\_id\_t
+
+\* @flags CREATE\_ONLY
+
+\* @objects SAI\_OBJECT\_TYPE\_PORT
+
+\*/
+
+SAI\_PORT\_CONNECTOR\_ATTR\_LINE\_SIDE\_FAILOVER\_PORT\_ID,
+
+
+## Additions to saiswitch.h
+
+/\*\*
+
+\* @brief Attribute data for #SAI\_SWITCH\_ATTR\_FAILOVER\_CONFIG\_MODE
+
+\* Used for Failover configuration mode
+
+\* In case of primary port failure, hitless enables the switch over to secondary
+
+\* port without losing link. It allows uninterrupted data transmission
+
+\*/
+
+typedef enum \_sai\_switch\_failover\_config\_mode\_t
+
+{
+
+/\*\* Ports are configured but do not operate in hitless \*/
+
+SAI\_SWITCH\_FAILOVER\_CONFIG\_MODE\_NO\_HITLESS,
+
+/\*\* Ports are configured and operate in hitless \*/
+
+SAI\_SWITCH\_FAILOVER\_CONFIG\_MODE\_HITLESS
+
+} sai\_switch\_failover\_config\_mode\_t;
+
+/\*\*
+
+\* @brief Failover configuration mode
+
+\*
+
+\* @type sai\_switch\_failover\_config\_mode\_t
+
+\* @flags CREATE\_AND\_SET
+
+\* @default SAI\_SWITCH\_FAILOVER\_CONFIG\_MODE\_NO\_HITLESS
+
+\*/
+
+SAI\_SWITCH\_ATTR\_FAILOVER\_CONFIG\_MODE,
+
+/\*\*
+
+\* @brief Query for Failover mode support
+
+\*
+
+\* @type bool
+
+\* @flags READ\_ONLY
+
+\*/
+
+SAI\_SWITCH\_ATTR\_SUPPORTED\_FAILOVER\_MODE,
+
+# Configuration Example
+
+Following example shows how to setup failover configuration on a system side port. In this example a 16 lane PHY is used.
+
+/\* Create System side, Line side, and System Side Failover ports \*/
```

Comment:
use code block?

Reply:
Configuration Example is in block and added below lines as bold